### PR TITLE
Refactor/net: split headers to proxy class

### PR DESCRIPTION
**Pull request by Ryex**
This PR splits the setting of headers out of the Net::Download class into a proxy class. This allows a more generic controll flow to use the download class without invoking a static cast on the QCoreApplication::inst() pointer 

(I'm trying to let things like the updater use the networks tasks)

### CODE REVIEW

**Comment by getchoo** [comment]
is this complete? just wondering since i noticed the old `Net::Download` functions weren't removed yet, and i'm not sure if that's for backwards compatibility. otherwise lgtm :)

**Comment by Ryex** [comment]
> is this complete? just wondering since i noticed the old `Net::Download` functions weren't removed yet, and i'm not sure if that's for backwards compatibility. otherwise lgtm :)

> is this complete? just wondering since i noticed the old `Net::Download` functions weren't removed yet, and i'm not sure if that's for backwards compatibility. otherwise lgtm :)

Ok this is fixed and finished.
Needed to add a compiler def to deal with the user agent and apply the changed to the `Upload` class too.

**Comment by Ryex** [comment]
> is this complete? just wondering since i noticed the old `Net::Download` functions weren't removed yet, and i'm not sure if that's for backwards compatibility. otherwise lgtm :)

the old functions are still used in some places and outside the the main launcher in the updater I'm building. which is the whole point of making this change.

**Comment by Trial97** [comment]
Maybe you can also unify Upload and Download as mostly the code base is the same.
Here is my idea: https://github.com/Ryex/PrismLauncher/compare/refactor/net-split-headers-to-proxy-class...Trial97:PrismLauncher:refactor/net-split-headers-to-proxy-class?expand=1

**Comment by Ryex** [comment]
> Maybe you can also unify Upload and Download as mostly the code base is the same. Here is my idea: https://github.com/Ryex/PrismLauncher/compare/refactor/net-split-headers-to-proxy-class...Trial97:PrismLauncher:refactor/net-split-headers-to-proxy-class?expand=1

Id actually very much like that refactor. can you clean up that patch though? there seems to be a bunch of extraneous stuff in it.

**Comment by Trial97** [comment]
> Id actually very much like that refactor. can you clean up that patch though? there seems to be a bunch of extraneous stuff in it.

Done https://github.com/Ryex/PrismLauncher/pull/1

**Comment by Ryex** [comment]
> > Id actually very much like that refactor. can you clean up that patch though? there seems to be a bunch of extraneous stuff in it.
> 
> Done [Ryex#1](https://github.com/Ryex/PrismLauncher/pull/1)

ty

**Review by getchoo — APPROVED** [review]
alright, previous issues look to be resolved; after running a few sets of downloads/updates from cf and modrinth, i haven't been able to find any regressions! the refactor also makes a lot of sense imo, along with the new netrequest class (and thanks trial for making that a separate pr, made it really easy to review for me :p)

adding this to the 8.0 milestone since a few others prs are dependent on it now

File: launcher/net/ApiDownload.cpp [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
```
Comment (TheKodeToad) [inline launcher/net/ApiDownload.cpp:18]: You should probably include MultiMC's license, as well as in other places (along with any applicable contributors)

File: launcher/net/ApiHeaderProxy.h [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    virtual ~ApiHeaderProxy() = default;
+
+   public:
+    virtual QList<HeaderPair> headers(const QNetworkRequest& request) const override
+    {
+        QList<HeaderPair> hdrs;
+        if (APPLICATION->capabilities() & Application::SupportsFlame && request.url().host() == QUrl(BuildConfig.FLAME_BASE_URL).host()) {
+            hdrs.append({ "x-api-key", APPLICATION->getFlameAPIKey().toUtf8() });
+        } else if (request.url().host() == QUrl(BuildConfig.MODRINTH_PROD_URL).host() ||
+                   request.url().host() == QUrl(BuildConfig.MODRINTH_STAGING_URL).host()) {
```
Comment (TheKodeToad) [inline launcher/net/ApiHeaderProxy.h:40]: (note) Should be moved in future

File: launcher/net/ApiHeaderProxy.h [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+#include "Application.h"
+#include "BuildConfig.h"
+#include "net/HeaderProxy.h"
+
+namespace Net {
+
+class ApiHeaderProxy : public HeaderProxy {
+   public:
+    ApiHeaderProxy() : HeaderProxy(){};
```
Comment (TheKodeToad) [inline launcher/net/ApiHeaderProxy.h]:
```suggestion
    ApiHeaderProxy() : HeaderProxy() {}
```

File: launcher/net/HeaderProxy.h [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+   public:
+    virtual QList<HeaderPair> headers(const QNetworkRequest& request) const = 0;
+
+   public:
+    void writeHeaders(QNetworkRequest& request)
+    {
+        for (auto header : headers(request)) {
+            request.setRawHeader(header.headerName, header.headerValue);
+        }
+    };
```
Comment (TheKodeToad) [inline launcher/net/HeaderProxy.h]:
```suggestion
    }
```

File: launcher/net/NetRequest.h [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+   public:
+    ~NetRequest() override = default;
+
+    void init() override{};
+
+   public:
+    void addValidator(Validator* v);
+    auto abort() -> bool override;
+    auto canAbort() const -> bool override { return true; };
```
Comment (TheKodeToad) [inline launcher/net/NetRequest.h]:
```suggestion
    auto canAbort() const -> bool override { return true; }
```
\>:(

File: launcher/net/RawHeaderProxy.h [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+#pragma once
+
+#include "net/HeaderProxy.h"
+
+namespace Net {
+
+class RawHeaderProxy : public HeaderProxy {
+   public:
+    RawHeaderProxy() : HeaderProxy(){};
```
Comment (TheKodeToad) [inline launcher/net/RawHeaderProxy.h]:
```suggestion
    RawHeaderProxy() : HeaderProxy() {}
```

File: launcher/net/Download.h [hunk 7]
```diff
@@ -38,57 +38,26 @@
...
     Q_OBJECT
-
    public:
     using Ptr = shared_qobject_ptr<class Download>;
-    enum class Option { NoOptions = 0, AcceptLocalFiles = 1, MakeEternal = 2 };
-    Q_DECLARE_FLAGS(Options, Option)
-
-   public:
-    ~Download() override = default;
+    explicit Download() : NetRequest() { logCat = taskDownloadLogC; };
```
Comment (TheKodeToad) [inline launcher/net/Download.h]:
```suggestion
    explicit Download() : NetRequest() { logCat = taskDownloadLogC; }
```

**Review by TheKodeToad — CHANGES_REQUESTED** [review]

File: launcher/net/ApiDownload.cpp (hunk 1 above)
Comment (Ryex) [reply launcher/net/ApiDownload.cpp:18]: Nah, the code in this fine is original. `NetRequest` now holds the reused code. which is why it has the referenced license.

Comment (TheKodeToad) [reply launcher/net/ApiDownload.cpp:18]: The `make*` functions look very similar to in Download.cpp

Comment (Ryex) [reply launcher/net/ApiDownload.cpp:18]: similar in name only. conforming to the same interface doesn't mean they aren't original.

Comment (TheKodeToad) [reply launcher/net/ApiDownload.cpp:18]: huh? the function implementations look almost the same

Comment (Ryex) [reply launcher/net/ApiDownload.cpp:18]: There are only so many ways to write them. They are new functions on a new class, the old functions still exist, there for these are original code. This is actually law in the US thanks to Oracle vs Google.

**Review by TheKodeToad — APPROVED** [review]

**Review by Scrumplex — APPROVED** [review]

**Comment by Ryex** [comment]
ok, forgiven.

**Comment by Scrumplex** [comment]
I don't trust my merge commits tbh. I used JetBrains' magic merge tool, and it might have done some questionable stuff. Let's just move it to staging instead

**Pull request merged**